File: runtime/flutter/cupertino/cupertinoContextMenuAction.ts
import { Type } from "../../dart/core/type";
import { VoidCallback } from "../../dart/ui/voidCallback";
import { RuntimeBaseClass } from "../../runtimeBaseClass";
import { Key } from "../foundation/key";
import { Widget } from "../widget";
import { IconData } from "../widgets/iconData";
import { StatelessWidget } from "../widgets/statelessWidget";

export interface CupertinoContextMenuActionProps {
    key?: Key | undefined;
    child: Widget;
    isDefaultAction?: boolean | undefined;
    isDestructiveAction?: boolean | undefined;
    onPressed?: VoidCallback | undefined;
    trailingIcon?: IconData | undefined;
}

declare const flutter: {
    cupertino: {
        cupertinoContextMenuAction: (
            this: void,
            props: CupertinoContextMenuActionProps
        ) => CupertinoContextMenuAction;
    };
};

export class CupertinoContextMenuAction
    extends StatelessWidget
    implements RuntimeBaseClass {
    public readonly internalRuntimeType = new Type(CupertinoContextMenuAction);
    public props: CupertinoContextMenuActionProps;
    public constructor(props: CupertinoContextMenuActionProps) {
        super();
        this.props = props;

        if (this.props.isDefaultAction === undefined) {
            this.props.isDefaultAction = false;
        }

        if (this.props.isDestructiveAction === undefined) {
            this.props.isDestructiveAction = false;
        }
    }

    public build() {
        return flutter.cupertino.cupertinoContextMenuAction(this.props);
    }
}

File: lib/cfr/builtins/libs/flutter/cupertino/cupertinoContextMenuAction.dart
import 'package:flutter/cupertino.dart';

import 'package:hydro_sdk/cfr/builtins/boxing/unboxers.dart';
import 'package:hydro_sdk/cfr/vm/closure.dart';
import 'package:hydro_sdk/cfr/vm/context.dart';
import 'package:hydro_sdk/cfr/vm/table.dart';
import 'package:hydro_sdk/hydroState.dart';

void loadCupertinoContextMenuAction(
    {required HydroState luaState, required HydroTable table}) {
  table["cupertinoContextMenuAction"] =
      makeLuaDartFunc(func: (List<dynamic> args) {
    return [
      CupertinoContextMenuAction(
        key: maybeUnBoxAndBuildArgument<Key, dynamic>(args[0]["key"],
            parentState: luaState),
        child: maybeUnBoxAndBuildArgument<Widget, dynamic>(args[0]["child"],
            parentState: luaState),
        isDefaultAction: args[0]["isDefaultAction"],
        isDestructiveAction: args[0]["isDestructiveAction"],
        onPressed: () {
          Closure? closure = args[0]["onPressed"];
          if (closure != null) {
            closure.dispatch(
              [],
              parentState: luaState,
              resetEnclosingLexicalEnvironment: true,
            );
          }
        },
        trailingIcon: maybeUnBoxAndBuildArgument<IconData, dynamic>(
            args[0]["trailingIcon"],
            parentState: luaState),
      )
    ];
  });
}
